File: src/test/java/kim/present/solve/programmers/Solution258712Test.java
package kim.present.solve.programmers;

import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

class Solution258712Test {

    static Solution258712 javaSolution = new Solution258712();
    static Solution258712kt kotlinSolution = new Solution258712kt();

    @Test
    void testcase_1() {
        String[] friends = {"muzi", "ryan", "frodo", "neo"};
        String[] gifts = {"muzi frodo", "muzi frodo", "ryan muzi", "ryan muzi", "ryan muzi", "frodo muzi", "frodo ryan", "neo muzi"};
        assertEquals(2, javaSolution.solution(friends, gifts));
        assertEquals(2, kotlinSolution.solution(friends, gifts));
    }

    @Test
    void testcase_2() {
        String[] friends = {"joy", "brad", "alessandro", "conan", "david"};
        String[] gifts = {"alessandro brad", "alessandro joy", "alessandro conan", "david alessandro", "alessandro david"};
        assertEquals(4, javaSolution.solution(friends, gifts));
        assertEquals(4, kotlinSolution.solution(friends, gifts));
    }

    @Test
    void testcase_3() {
        String[] friends = {"a", "b", "c"};
        String[] gifts = {"a b", "b a", "c a", "a c", "a c", "c a"};
        assertEquals(0, javaSolution.solution(friends, gifts));
        assertEquals(0, kotlinSolution.solution(friends, gifts));
    }
}

File: src/test/java/kim/present/solve/programmers/Solution258712Test.kt
package kim.present.solve.programmers

import org.junit.jupiter.api.TestFactory

internal class Solution258712Test : TestBase2<Array<String>, Array<String>, Int>() {
    init {
        +Solution258712()::solution
        +Solution258712kt()::solution
    }

    @TestFactory
    fun run() = test {
        {
            of("muzi", "ryan", "frodo", "neo") args of(
                "muzi frodo",
                "muzi frodo",
                "ryan muzi",
                "ryan muzi",
                "ryan muzi",
                "frodo muzi",
                "frodo ryan",
                "neo muzi"
            )
        } expected 2

        {
            of("joy", "brad", "alessandro", "conan", "david") args of(
                "alessandro brad",
                "alessandro joy",
                "alessandro conan",
                "david alessandro",
                "alessandro david"
            )
        } expected 4

        {
            of("a", "b", "c") args of(
                "a b",
                "b a",
                "c a",
                "a c",
                "a c",
                "c a"
            )
        } expected 0
    }
}
